cpp/src/arrow/compute/kernels/scalar_cast_test.cc

```diff
@@ -2245,10 +2245,132 @@ static void CheckStructToStruct(
+          auto dest7, StructArray::Make({a2, b2, c2, d2, e2}, {"a", "b", "c", "d", "e"}));
+      CheckCast(src, dest7);
+
+      const auto dest8 = arrow::struct_({std::make_shared<Field>("a", int8()),
+                                         std::make_shared<Field>("d", int16()),
+                                         std::make_shared<Field>("f", int64())});
+      const auto options = CastOptions::Safe(dest8);
+      EXPECT_RAISES_WITH_MESSAGE_THAT(
+          TypeError,
+          ::testing::HasSubstr(
+              "struct (sub)fields don't match or are in the wrong order"),
+          Cast(src, options));
+
+      // With nulls
+      std::shared_ptr<Buffer> null_bitmap;
+      BitmapFromVector<int>({0, 1, 0}, &null_bitmap);
+
+      ASSERT_OK_AND_ASSIGN(auto src_null, StructArray::Make({a1, b1, c1, d1, e1},
+                                                            field_names, null_bitmap));
+      ASSERT_OK_AND_ASSIGN(
+          auto dest1_null,
+          StructArray::Make({a2, c2}, std::vector<std::string>{"a", "c"}, null_bitmap));
+      CheckCast(src_null, dest1_null);
+
+      ASSERT_OK_AND_ASSIGN(
+          auto dest2_null,
+          StructArray::Make({b2, d2}, std::vector<std::string>{"b", "d"}, null_bitmap));
+      CheckCast(src_null, dest2_null);
+
+      ASSERT_OK_AND_ASSIGN(
+          auto dest3_null,
+          StructArray::Make({c2, e2}, std::vector<std::string>{"c", "e"}, null_bitmap));
+      CheckCast(src_null, dest3_null);
+
+      ASSERT_OK_AND_ASSIGN(
+          auto dest4_null,
+          StructArray::Make({a2, d2, e2}, std::vector<std::string>{"a", "d", "e"},
+                            null_bitmap));
+      CheckCast(src_null, dest4_null);
+
+      ASSERT_OK_AND_ASSIGN(
+          auto dest5_null,
+          StructArray::Make({b2, c2, e2}, std::vector<std::string>{"b", "c", "e"},
+                            null_bitmap));
+      CheckCast(src_null, dest5_null);
+
+      ASSERT_OK_AND_ASSIGN(
+          auto dest6_null,
+          StructArray::Make({a2, b2, c2, e2},
+                            std::vector<std::string>{"a", "b", "c", "e"}, null_bitmap));
+      CheckCast(src_null, dest6_null);
+
+      ASSERT_OK_AND_ASSIGN(
+          auto dest7_null,
+          StructArray::Make({a2, b2, c2, d2, e2},
+                            std::vector<std::string>{"a", "b", "c", "d", "e"},
+                            null_bitmap));
+      CheckCast(src_null, dest7_null);
+
+      const auto dest8_null = arrow::struct_({std::make_shared<Field>("a", int8()),
+                                              std::make_shared<Field>("d", int16()),
+                                              std::make_shared<Field>("f", int64())});
+      const auto options_null = CastOptions::Safe(dest8_null);
+      EXPECT_RAISES_WITH_MESSAGE_THAT(
+          TypeError,
+          ::testing::HasSubstr(
+              "struct (sub)fields don't match or are in the wrong order"),
+          Cast(src_null, options_null));
```

Comment:
Can we also add a case or two dealing with duplicate field names on either side? (It looks like the implementation will handle this sensibly.)